File: genui/adapters/test_adapter.py
# -*- coding: utf-8 -*-
"""测试适配器, 支持自动化测试而不需要真实UI"""

from typing import Any, Dict, Callable, List, Tuple, Optional
from genui.adapters.base import AdapterBase
from genui.core.component import Component
from genui.core.ui_instance import UIInstance


class VirtualWidget:
    """虚拟组件, 模拟真实UI组件"""

    def __init__(self, component: Component):
        """初始化虚拟组件

        Args:
            component: 抽象组件定义
        """
        self.component = component
        self.id = component.id
        self.type = component.type
        self.value: Any = None
        self.event_handlers: Dict[str, Callable] = {}

        # 根据组件类型初始化默认值
        if component.type == "text_input":
            self.value = getattr(component, "default_value", "")
        elif component.type == "checkbox":
            self.value = getattr(component, "checked", False)
        elif component.type == "radio_group":
            self.value = getattr(component, "selected", None)
        elif component.type == "dropdown":
            self.value = getattr(component, "selected", None)
        elif component.type == "label":
            self.value = getattr(component, "text", "")

    def get(self) -> Any:
        """获取组件值 (模拟Tkinter的get方法)"""
        return self.value

    def set(self, value: Any):
        """设置组件值"""
        self.value = value

    def config(self, **kwargs):
        """配置组件属性 (模拟Tkinter的config方法)"""
        for key, value in kwargs.items():
            if key == "text":
                self.value = value


class TestAdapter(AdapterBase):
    """测试适配器, 在内存中模拟UI交互"""

    def __init__(self):
        """初始化测试适配器"""
        self.widgets: Dict[str, VirtualWidget] = {}  # 存储所有虚拟组件
        self.events: List[Tuple[str, str]] = []  # 记录触发的事件 (event_type, component_id)
        self.outputs: List[str] = []  # 记录 display() 输出
        self.event_handlers: Dict[str, Callable] = {}  # 事件处理函数
        self.tool_executor: Optional[Any] = None  # Tool执行器

    def render_component(
        self,
        component: Component,
        parent: Any,
        event_handlers: Dict[str, Callable]
    ) -> VirtualWidget:
        """渲染组件到内存

        Args:
            component: 要渲染的抽象组件
            parent: 父组件
            event_handlers: 事件处理函数字典

        Returns:
            虚拟组件实例
        """
        # 创建虚拟组件
        widget = VirtualWidget(component)
        self.widgets[component.id] = widget

        # 绑定事件处理器
        if component.type == "button":
            on_click = getattr(component, "on_click", None)
            if on_click and on_click in event_handlers:
                widget.event_handlers["click"] = event_handlers[on_click]

        elif component.type == "checkbox":
            on_change = getattr(component, "on_change", None)
            if on_change and on_change in event_handlers:
                widget.event_handlers["change"] = event_handlers[on_change]

        elif component.type == "radio_group":
            on_change = getattr(component, "on_change", None)
            if on_change and on_change in event_handlers:
                widget.event_handlers["change"] = event_handlers[on_change]

        # 递归渲染子组件
        if component.type == "container":
            children = getattr(component, "children", [])
            for child in children:
                self.render_component(child, widget, event_handlers)

        return widget

    def create_window(
        self,
        ui_instance: UIInstance,
        tool_executor: Any = None
    ) -> Dict[str, Any]:
        """创建虚拟窗口

        Args:
            ui_instance: UI实例
            tool_executor: Tool执行器

        Returns:
            虚拟窗口对象 (字典)
        """
        self.tool_executor = tool_executor

        # 构建辅助函数环境
        helper_env = {
            "get_widget": self._get_widget,
            "get_value": self._get_value,
            "set_value": self._set_value,
            "update_widget": self._update_widget,
            "display": self._display,
            "print": print,
        }

        # 如果有 tool_executor, 添加 call_function
        if tool_executor:
            helper_env["call_function"] = tool_executor.call_function

        # 编译事件处理函数
        compiled_handlers = {}
        for name, code in ui_instance.event_handlers.items():
            try:
                exec(code, helper_env)
                compiled_handlers[name] = helper_env[name]
            except Exception as e:
                raise RuntimeError(f"编译事件处理函数 '{name}' 失败: {e}")

        self.event_handlers = compiled_handlers

        # 渲染根组件
        root_widget = self.render_component(
            ui_instance.root,
            None,
            compiled_handlers
        )

        return {
            "title": ui_instance.title,
            "width": ui_instance.width,
            "height": ui_instance.height,
            "root": root_widget
        }

    def run_event_loop(self, window: Any) -> None:
        """不运行真实事件循环 (测试模式下不需要)

        Args:
            window: 虚拟窗口对象
        """
        pass

    # ========== 测试辅助方法 ==========

    def click(self, component_id: str):
        """模拟点击按钮

        Args:
            component_id: 组件ID

        Raises:
            ValueError: 如果组件不存在或不是按钮
        """
        widget = self.widgets.get(component_id)
        if not widget:
            raise ValueError(f"组件 '{component_id}' 不存在")

        if widget.type != "button":
            raise ValueError(f"组件 '{component_id}' 不是按钮")

        # 触发点击事件
        handler = widget.event_handlers.get("click")
        if handler:
            handler()
            self.events.append(("click", component_id))

    def set_input(self, component_id: str, value: str):
        """模拟输入文本

        Args:
            component_id: 组件ID
            value: 输入值

        Raises:
            ValueError: 如果组件不存在或不支持输入
        """
        widget = self.widgets.get(component_id)
        if not widget:
            raise ValueError(f"组件 '{component_id}' 不存在")

        if widget.type not in ["text_input"]:
            raise ValueError(f"组件 '{component_id}' 不支持输入")

        widget.value = value

    def get_output(self, component_id: str) -> Any:
        """获取组件输出值

        Args:
            component_id: 组件ID

        Returns:
            组件的当前值

        Raises:
            ValueError: 如果组件不存在
        """
        widget = self.widgets.get(component_id)
        if not widget:
            raise ValueError(f"组件 '{component_id}' 不存在")

        return widget.value

    def select(self, component_id: str, option: str):
        """模拟选择选项 (下拉框或单选按钮组)

        Args:
            component_id: 组件ID
            option: 选择的选项

        Raises:
            ValueError: 如果组件不存在或不支持选择
        """
        widget = self.widgets.get(component_id)
        if not widget:
            raise ValueError(f"组件 '{component_id}' 不存在")

        if widget.type not in ["dropdown", "radio_group"]:
            raise ValueError(f"组件 '{component_id}' 不支持选择")

        widget.value = option

        # 触发 change 事件
        handler = widget.event_handlers.get("change")
        if handler:
            handler()
            self.events.append(("change", component_id))

    def check(self, component_id: str, checked: bool = True):
        """模拟勾选/取消复选框

        Args:
            component_id: 组件ID
            checked: 是否勾选

        Raises:
            ValueError: 如果组件不存在或不是复选框
        """
        widget = self.widgets.get(component_id)
        if not widget:
            raise ValueError(f"组件 '{component_id}' 不存在")

        if widget.type != "checkbox":
            raise ValueError(f"组件 '{component_id}' 不是复选框")

        widget.value = checked

        # 触发 change 事件
        handler = widget.event_handlers.get("change")
        if handler:
            handler()
            self.events.append(("change", component_id))

    def get_events(self) -> List[Tuple[str, str]]:
        """获取所有触发的事件

        Returns:
            事件列表 [(event_type, component_id), ...]
        """
        return self.events.copy()

    def get_outputs(self) -> List[str]:
        """获取所有 display() 输出

        Returns:
            输出列表
        """
        return self.outputs.copy()

    def clear_events(self):
        """清空事件记录"""
        self.events.clear()

    def clear_outputs(self):
        """清空输出记录"""
        self.outputs.clear()

    # ========== 内部辅助函数 (供事件处理器使用) ==========

    def _get_widget(self, component_id: str) -> VirtualWidget:
        """获取虚拟组件"""
        widget = self.widgets.get(component_id)
        if not widget:
            raise ValueError(f"组件 '{component_id}' 不存在")
        return widget

    def _get_value(self, component_id: str) -> Any:
        """获取组件值"""
        return self.get_output(component_id)

    def _set_value(self, component_id: str, value: Any):
        """设置组件值"""
        widget = self._get_widget(component_id)
        widget.value = value

    def _update_widget(self, component_id: str, **kwargs):
        """更新组件属性"""
        widget = self._get_widget(component_id)
        for key, value in kwargs.items():
            if key == "text":
                widget.value = value

    def _display(self, message: str):
        """记录输出消息"""
        self.outputs.append(message)

File: src/adapters/test-adapter.ts
// -*- coding: utf-8 -*-
// 测试适配器, 提供无头(headless)环境用于单元测试UI交互

import type { UIAdapter } from './base.js';
import type { UIInstance } from '../core/ui-instance.js';
import type { ComponentType } from '../core/component.js';
import { listAllComponents } from '../core/ui-instance.js';
import { createLogger } from '../logger.js';

const logger = createLogger('TestAdapter');

/**
 * 虚拟组件, 用于在无头环境中模拟UI组件的状态
 */
export interface VirtualWidget {
  /** 组件ID */
  id: string;
  /** 组件类型 */
  type: string;
  /** 组件当前值 */
  value: unknown;
  /** 对应的原始组件定义 */
  component: ComponentType;
  /** 事件处理器映射(事件名 -> 处理函数) */
  eventHandlers: Map<string, () => void>;
}

/**
 * 测试适配器, 用于在无头环境中模拟UI交互.
 *
 * 功能:
 * - 将UI实例的组件渲染为VirtualWidget
 * - 编译事件处理器字符串为可执行函数
 * - 提供模拟用户操作的方法(click, setInput, select, check等)
 * - 记录事件和display输出, 方便测试验证
 */
export class TestAdapter implements UIAdapter {
  readonly name: string = 'test';

  /** 虚拟组件存储(id -> VirtualWidget) */
  private widgets: Map<string, VirtualWidget> = new Map();

  /** 记录的事件列表, 每个事件为[事件类型, 组件ID] */
  private events: Array<[string, string]> = [];

  /** display()调用的输出记录 */
  private displayOutputs: string[] = [];

  /** 编译后的事件处理器函数(处理器名 -> 函数) */
  private compiledHandlers: Map<string, () => void> = new Map();

  async createWindow(uiInstance: UIInstance, _toolExecutor?: any): Promise<void> {
    logger.info(`创建测试窗口: ${uiInstance.title}`);

    // 重置状态
    this.widgets.clear();
    this.events = [];
    this.displayOutputs = [];
    this.compiledHandlers.clear();

    // 编译事件处理器
    this.compileEventHandlers(uiInstance.event_handlers);

    // 递归渲染所有组件
    const allComponents = listAllComponents(uiInstance);
    for (const comp of allComponents) {
      this.renderComponent(comp);
    }

    // 绑定事件
    this.bindEvents(allComponents);

    logger.info(`渲染完成, 共${this.widgets.size}个组件`);
  }

  async runEventLoop(): Promise<void> {
    // 测试适配器不需要事件循环, 立即返回
  }

  // ============================================================
  // 测试辅助方法
  // ============================================================

  /**
   * 模拟点击按钮.
   *
   * @param id - 按钮组件ID
   */
  click(id: string): void {
    const widget = this.widgets.get(id);
    if (!widget) {
      throw new Error(`组件不存在: ${id}`);
    }

    this.events.push(['click', id]);

    const handler = widget.eventHandlers.get('click');
    if (handler) {
      handler();
    }
  }

  /**
   * 模拟设置文本输入框的值.
   *
   * @param id - 文本输入框组件ID
   * @param value - 要设置的值
   */
  setInput(id: string, value: string): void {
    const widget = this.widgets.get(id);
    if (!widget) {
      throw new Error(`组件不存在: ${id}`);
    }
    widget.value = value;
  }

  /**
   * 获取组件的当前值.
   *
   * @param id - 组件ID
   * @returns 组件的当前值
   */
  getOutput(id: string): unknown {
    const widget = this.widgets.get(id);
    if (!widget) {
      throw new Error(`组件不存在: ${id}`);
    }
    return widget.value;
  }

  /**
   * 模拟选择dropdown或radio_group的选项.
   *
   * @param id - 组件ID
   * @param option - 选中的选项值
   */
  select(id: string, option: string): void {
    const widget = this.widgets.get(id);
    if (!widget) {
      throw new Error(`组件不存在: ${id}`);
    }
    widget.value = option;

    this.events.push(['change', id]);

    const handler = widget.eventHandlers.get('change');
    if (handler) {
      handler();
    }
  }

  /**
   * 模拟勾选/取消勾选checkbox.
   *
   * @param id - checkbox组件ID
   * @param checked - 是否勾选
   */
  check(id: string, checked: boolean): void {
    const widget = this.widgets.get(id);
    if (!widget) {
      throw new Error(`组件不存在: ${id}`);
    }
    widget.value = checked;

    this.events.push(['change', id]);

    const handler = widget.eventHandlers.get('change');
    if (handler) {
      handler();
    }
  }

  /**
   * 获取VirtualWidget.
   *
   * @param id - 组件ID
   * @returns VirtualWidget或undefined
   */
  getWidget(id: string): VirtualWidget | undefined {
    return this.widgets.get(id);
  }

  /**
   * 获取记录的事件列表.
   *
   * @returns 事件列表, 每个元素为[事件类型, 组件ID]
   */
  getEvents(): Array<[string, string]> {
    return [...this.events];
  }

  /**
   * 获取display()调用输出的消息列表.
   *
   * @returns display输出消息列表
   */
  getDisplayOutputs(): string[] {
    return [...this.displayOutputs];
  }

  /**
   * 清除记录的事件.
   */
  clearEvents(): void {
    this.events = [];
  }

  /**
   * 清除display输出记录.
   */
  clearOutputs(): void {
    this.displayOutputs = [];
  }

  // ============================================================
  // 私有方法
  // ============================================================

  /**
   * 根据组件类型获取初始值.
   *
   * @param component - 组件定义
   * @returns 初始值
   */
  private getInitialValue(component: ComponentType): unknown {
    switch (component.type) {
      case 'text_input':
        return component.default_value || '';
      case 'checkbox':
        return component.checked ?? false;
      case 'radio_group':
        return component.selected ?? null;
      case 'dropdown':
        return component.selected ?? null;
      case 'label':
        return component.text || '';
      default:
        return null;
    }
  }

  /**
   * 渲染单个组件为VirtualWidget.
   *
   * @param component - 组件定义
   */
  private renderComponent(component: ComponentType): void {
    const widget: VirtualWidget = {
      id: component.id,
      type: component.type,
      value: this.getInitialValue(component),
      component,
      eventHandlers: new Map(),
    };
    this.widgets.set(component.id, widget);
  }

  /**
   * 编译事件处理器字符串为可执行函数.
   * 处理器代码中可以使用以下辅助函数:
   * - getValue(id) - 获取组件值
   * - setValue(id, value) - 设置组件值
   * - getWidget(id) - 获取虚拟组件
   * - updateWidget(id, props) - 更新组件属性
   * - display(message) - 显示消息(记录到输出)
   * - callFunction(name, ...args) - 调用其他处理器
   *
   * @param handlers - 事件处理器名->代码字符串映射
   */
  private compileEventHandlers(handlers: Record<string, string>): void {
    // 构建辅助函数环境
    const envKeys = [
      'getValue',
      'setValue',
      'getWidget',
      'updateWidget',
      'display',
      'callFunction',
    ];

    const envValues = [
      // getValue
      (id: string): unknown => {
        const w = this.widgets.get(id);
        if (!w) {
          throw new Error(`getValue: 组件不存在: ${id}`);
        }
        return w.value;
      },
      // setValue
      (id: string, value: unknown): void => {
        const w = this.widgets.get(id);
        if (!w) {
          throw new Error(`setValue: 组件不存在: ${id}`);
        }
        w.value = value;
      },
      // getWidget
      (id: string): VirtualWidget | undefined => {
        return this.widgets.get(id);
      },
      // updateWidget
      (id: string, props: Record<string, unknown>): void => {
        const w = this.widgets.get(id);
        if (!w) {
          throw new Error(`updateWidget: 组件不存在: ${id}`);
        }
        Object.assign(w, props);
      },
      // display
      (message: string): void => {
        this.displayOutputs.push(message);
        logger.info(`display: ${message}`);
      },
      // callFunction
      (name: string, ...args: unknown[]): unknown => {
        const fn = this.compiledHandlers.get(name);
        if (!fn) {
          throw new Error(`callFunction: 处理器不存在: ${name}`);
        }
        return (fn as (...args: unknown[]) => unknown)(...args);
      },
    ];

    for (const [handlerName, code] of Object.entries(handlers)) {
      try {
        // 使用new Function编译事件处理器
        // 在代码末尾添加return语句, 返回命名函数
        const fn = new Function(
          ...envKeys,
          `${code}\nreturn ${handlerName};`,
        );
        const handler = fn(...envValues);
        this.compiledHandlers.set(handlerName, handler);
      } catch (err) {
        logger.error(`编译事件处理器失败: ${handlerName}`, err);
      }
    }
  }

  /**
   * 为组件绑定事件处理器.
   *
   * @param components - 所有组件列表
   */
  private bindEvents(components: ComponentType[]): void {
    for (const comp of components) {
      const widget = this.widgets.get(comp.id);
      if (!widget) {
        continue;
      }

      // button的on_click -> click事件
      if (comp.type === 'button' && 'on_click' in comp && comp.on_click) {
        const handler = this.compiledHandlers.get(comp.on_click);
        if (handler) {
          widget.eventHandlers.set('click', handler);
        }
      }

      // checkbox的on_change -> change事件
      if (comp.type === 'checkbox' && 'on_change' in comp && comp.on_change) {
        const handler = this.compiledHandlers.get(comp.on_change);
        if (handler) {
          widget.eventHandlers.set('change', handler);
        }
      }

      // radio_group的on_change -> change事件
      if (comp.type === 'radio_group' && 'on_change' in comp && comp.on_change) {
        const handler = this.compiledHandlers.get(comp.on_change);
        if (handler) {
          widget.eventHandlers.set('change', handler);
        }
      }
    }
  }
}
